graftm/tree_decorator.py

```diff
@@ -1,309 +1,187 @@
+                                                               unique_number))
+                            seen_taxonomy_rank[classification]+=1
                         else:
-                            current_index = len(tax_list)
-                
-                if resolution < MAX_RESOLUTION:
-                    new_tax_list = [] 
+                            tax_string_array.append(classification) 
+                            seen_taxonomy_rank[classification] = 1
+                            
+            tax_string_array = [tax for tax in tax_string_array
+                                if tax not in self.GREENGENES_PREFIXES]
+            
+            if any(tax_string_array):
+                index = 0
+                for anc in node.ancestors():
                     
-                        
-                    for tax in tax_list:
-                        if not no_unique_names:
-                            if tax in self.encountered_taxonomies:
-                                idx = 1
-                                while tax in self.encountered_taxonomies:
-                                    if idx > 1: 
-                                        tax = '%s_%i' % ('_'.join(tax.split('_')[:-1]), idx)
-                                    else:
-                                        tax = tax + '_%i' % idx
-                                    idx += 1
+                    if hasattr(anc, 'index'):
+                        index+=anc.index
+                        if index >= self.max_taxonomy_index:break
 
-                        new_tax_list.append(tax)
-                        self.encountered_taxonomies.add(tax)
-                    tax_string = '; '.join(new_tax_list)
-                    
-                    if node.name:
-                        split_node_name = node.name.split(':')
-                        if len(split_node_name) > 2: 
-                            raise Exception("Malformed tree. Please remove ':' from the read names in the tree")
-                            exit(1)
-                        elif len(split_node_name) < 2:
-                            node_name = "%s:%s" % (split_node_name[0], tax_string)
-                        else:
-                            bootstrap_value, node_name = split_node_name[0], split_node_name[1]
-                            node_name = "%s:%s" % (bootstrap_value, tax_string)
-                    else:
-                        node_name = tax_string
-                    node.name = node_name
-                    logging.debug("Renamed node %s" % node_name)
-                    self.encountered_nodes[node]=index
+                tax_string_array = tax_string_array[index:]
+                
+                if len(tax_string_array) >0:
+                    node.name =  Taxonomy.GG_SEP_0.join(tax_string_array)
+                    node.index = len(tax_string_array)
+        
+    def return_tree(self):
```

Comment:
If more functions are added to tree_decorator (which I intended). This allows me to call functions on the tree (specified by the user and interpreted through decorator.main) until I need the end result i.e. when the tree is written to file. That way I wont be playing ping pong with the tree whenever I need to decorate it with something else.